### PR TITLE
Add ability to expand or collapse all the children of selected node

### PR DESCRIPTION
**Pull request by shriah**
Adds #591 

Took inspiration from `_revealDeep`. Not sure how efficient it is when the total nodes are large. Added shortcuts in `keyboardNav` but not sure if it is the right way to do it. This should work in all the shells.

### CODE REVIEW

File: frontend/keyboardNav.js [hunk 1]
```diff
@@ -26,6 +26,9 @@ var keyCodes = {
   '38': 'up',
   '39': 'right',
   '40': 'down',
+
+  '69': 'expand',
```
Comment (gaearon) [inline frontend/keyboardNav.js]: Please change this to match Chrome shortcuts.

File: frontend/Store.js [hunk 2]
```diff
@@ -340,6 +340,14 @@ class Store extends EventEmitter {
     this.emit(id);
   }
 
+  toggle(value: boolean) {
```
Comment (gaearon) [inline frontend/Store.js]: Can we give this a more descriptive name?

Comment (shriah) [inline frontend/Store.js]: will `toggleAllChildrenNode` be more descriptive ?

File: frontend/keyboardNav.js (hunk 1 above)
Comment (shriah) [inline frontend/keyboardNav.js]: I will have to add the shortcut to manifest.json of shells/chrome. Right now there is no support for shortcut in any of the plugins. Will spend some time to try to figure out how to do this.

File: frontend/Store.js (hunk 2 above)
Comment (gaearon) [inline frontend/Store.js]: Yea. ("Nodes" if plural.)

File: frontend/keyboardNav.js (hunk 1 above)
Comment (shriah) [inline frontend/keyboardNav.js]: I added support for using chrome shortcuts by adding it in the manifest file but the shortcut stopped working in the chrome elements panel. Not sure if this is desired.

I also tried this feature in a recently large react app and the expand all on the root node has a small lag and the close all works well. This lag is there even in the elements panel of chrome so this should be acceptable.

**Comment by shriah** [comment]
Updated the PR with support for chrome shortcuts.

File: frontend/keyboardNav.js [hunk 3]
```diff
@@ -43,6 +43,17 @@ module.exports = function keyboardNav(store: Store, win: Object): (e: DOMEvent)
       return;
     }
     e.preventDefault();
+    if ((e.altKey && direction === 'right') || (e.ctrlKey && e.altKey && direction === 'right')) {
```
Comment (gaearon) [inline frontend/keyboardNav.js]: Isn't the second condition extraneous? It seems to me that the left is always true when the right is true.
Same below.

File: frontend/Store.js [hunk 4]
```diff
@@ -535,6 +544,22 @@ class Store extends EventEmitter {
...
+    }
+    if (!(node.get('collapsed') === value)) {
+      this._nodes = this._nodes.setIn([id, 'collapsed'], value);
+      this.emit(id);
+    }
+    var children = node.get('children');
+    if (children && children.forEach) {
+      children.forEach(cid => this._toggleDeepChildren(cid, value));
+    }
+
```
Comment (gaearon) [inline frontend/Store.js]: Nit: unnecessary newline.

File: frontend/keyboardNav.js (hunk 3 above)
Comment (shriah) [inline frontend/keyboardNav.js]: I added this to disallow Control + Right Arrow

Comment (gaearon) [inline frontend/keyboardNav.js]: `if (e.altKey && direction === 'right')` is enough to match both Alt+Right and Ctrl+Alt+Right, isn't it? And it wouldn't match Ctrl+Right.

**Comment by gaearon** [comment]
This looks good, thanks!

**Comment by aspiers** [comment]
Cool, thanks a lot! What is the shortcut on Linux?

**Comment by gaearon** [comment]
Alt+Right and Alt+Left should work.

**Comment by gaearon** [comment]
(But the change is not released yet.)

**Comment by zhanwenchen** [comment]
Is this released yet? It's been 4 months

**Comment by bvaughn** [comment]
Yes 😄  Probably faster to just try it out than post 😁

**Comment by zhanwenchen** [comment]
Ah it's Alt+Left_Arrow and Alt+Right_Arrow

**Pull request merged**